share/fsmon/monitor.go

```diff
@@ -588,9 +575,62 @@ func (w *FileWatch) addDir(bIncInotify bool, finfo *osutil.FileInfoExt, files ma
 	}
 }
 
-func (w *FileWatch) getDirAndFileList(pid int, path, regx, cid string, filter *filterRegex, recur, protect, userAdded bool,
-	dirList map[string]*osutil.FileInfoExt) []*osutil.FileInfoExt {
-	dirs, singles := w.getDirFileList(pid, path, regx, cid, filter, recur, protect, userAdded)
+func getBaseDirPrefix(filter share.CLUSFileMonitorFilter) string {
+	dir := strings.Replace(filter.Path, "\\.", ".", -1)
+	if index := strings.Index(dir, ".*"); index > 0 {
+		dir = dir[:(index - 1)]
+	}
+	if filter.Regex == "" {
+		dir = filepath.Dir(dir)
+	}
+	return dir
+}
+
+func (w *FileWatch) getRootFs(pid int, cid string, profile *share.CLUSFileMonitorProfile) *workerlet.WalkPathResult {
+	var dirs []string
+	if profile != nil {
+		dirSet := utils.NewSet()
+		for _, filter := range profile.Filters {
+			dirSet.Add(getBaseDirPrefix(filter))
+		}
+		for _, filter := range profile.FiltersCRD {
+			dirSet.Add(getBaseDirPrefix(filter))
+		}
+		dirs = dirSet.ToStringSlice()
+		if pid > 1 {
+			log.WithFields(log.Fields{"dirs": dirs}).Debug("JW:")
+		}
+	}
+
+	res := workerlet.WalkPathResult{
+		Dirs:  make([]*workerlet.DirData, 0),
+		Files: make([]*workerlet.FileData, 0),
+	}
+
+	req := workerlet.WalkPathRequest{
+		Pid:     pid,
+		Path:    "/",
+		Timeout: rootIterTimeout,
+		Dirs:    dirs,
+	}
+
+	var err error
+	if err = w.walkerLimiter.Acquire(context.Background(), 1); err == nil {
```

Comment:
According to [docs](https://pkg.go.dev/golang.org/x/sync/semaphore#Weighted.Acquire), if a walker takes a lot of time or gets stuck, this go routine could wait forever in this `semaphore.Acquire()` call.  

Please utilize a timeout or use cancel(), so in the worst case it can give up and show error message to us.  

```suggestion
        ctx, cancel := context.WithTimeout(context.Background, req.Timeout)
        defer cancel()
	if err = w.walkerLimiter.Acquire(ctx, 1); err == nil {
```

Reply:
All tasks need to execute and can not be failed calls. The timeout is already applied on the worker.

Reply:
If it's expected that there shouldn't be timeout when acquiring a semaphore, could you give it a long timeout, e.g., 1 hour, so in the rare case it happens, we can still see the error?

Reply:
As I mentioned, the timeouts were applied at their following workers. The semaphore timeout is not required there.

Reply:
For sure.  Any concern if we change it like this?  We should be very careful about every lock mechanism because when it happens, it's very difficult to debug.  I don't think printing an error when it times out would be harmful.  Any different thoughts? 

```suggestion
        ctx, cancel := context.WithTimeout(context.Background, time.Hour * 1)
        defer cancel()
	if err = w.walkerLimiter.Acquire(ctx, 1); err == nil {
```

Reply:
Please read further the codes before make a suggestion. It is a double effort and will make disaster damages since there is not a failure handler at the other sides because those requests are must-executed operations.

IF you do think the extra timeout is required, please elaborate a scenario which could cause a deadlock.

Reply:
Every syscall could hang if there is a kernel hooking mechanism from a competitor or a kernel bug.  Every I/O call can hang if there is an I/O issue.  There are just too many possibilities that these calls could fail, that's why we want more logs for the unlikely.  The worst errors are those silent ones.

Reply:
Having said that, I understand your point - the code will work in most of time.  I just don't quite understand adding an error handling using context could be disastrous.  Could you elaborate more about it?

Reply:
Discussed offline and we agree this is not blocking this PR.